### ISSUE TITLE
Error "runtime error: index out of range"

### ISSUE THREAD

**Issue by allada** [binance-chain/bsc-erigon#9]
FYI: I just started syncing a new node and got this error in the console output:
```
INFO[12-21|09:03:31.699] [1/16 Headers] Inserting headers         progress=4227129
INFO[12-21|09:04:01.699] [1/16 Headers] Inserting headers         progress=4335429
INFO[12-21|09:04:31.699] [1/16 Headers] Inserting headers         progress=4462916
EROR[12-21|09:04:57.112] Staged Sync                              error="runtime error: index out of range [125] with length 125, trace: [stageloop.go:99 panic.go:1052 panic.go:90 header_data_struct.go:116 heap.go:93 heap.go:86 header_algos.go:594 stage_headers.go:392 stage_headers.go:124 default_stages.go:20 sync.go:318 sync.go:235 stageloop.go:133 stageloop.go:59 asm_arm64.s:1133]"
INFO[12-21|09:04:57.629] [1/16 Headers] Waiting for headers...    from=2251776
INFO[12-21|09:05:27.629] [1/16 Headers] Inserting headers         progress=2339773
```

It is still running, but figured I'd give a heads up... Not sure if this is a bug due to the modifications for BSC chain or erigon itself, but figured I'd start here.

**Comment by havsar** [comment]
@allada I had the same, but syncing continued afterwards correctly.

**Comment by allada** [comment]
FYI, I found that I had to start and stop it frequently (ended up writing a simple script) so it would save the state on shutdown, and restore during startup, otherwise it would start over from beginning every time it would happen.

**Comment by havsar** [comment]
@allada Do you mind sharing the script?

**Comment by allada** [comment]
I only needed to run this for the first stage "`Headers`", after this stage you can (and should) just run it normally.

```sh
#!/bin/bash
for i in {0..100000}
do
  timeout 10m erigon --datadir data --chain bsc --nat extip:put-ip-address-here
done
```
You may not need all those flags, but I found that the external IP discovery did not work, so I put it in manually.

**Comment by allada** [comment]
FYI, If you want to save a day or so of time, I'm in the process of building this database, but you can resume the data from this snapshot, here's about 1TB of it so far:
`s3://public-blockchain-snapshots/bsc/erigon-partial-2021-12-21.tar.zstd`

You'll need an aws account yourself to download, but it is public. It is configured for "requester pays". If you download outside of AWS's `us-west-2` the AWS s3 transfer costs will be about $25USD (so be warned).

Raw aws command is:
```
cd /place_to_put_data
aws s3 cp --request-payer=requester s3://public-blockchain-snapshots/bsc/erigon-partial-2021-12-21.tar.zstd - | pv | zstd -d --long=31 | tar xf -
```

**Comment by havsar** [comment]
@allada Did you manage to fully sync up? I'm at 6/16 execution stage but it looks like this can take ages. Currently at block 5m processing ~30 blocks/s.

**Comment by allada** [comment]
I estimate it'll take about 3 weeks to get a full archive node (just guesstimate), so don't expect any magic here.

I will be hosting this snapshot officially when finished. If you need an archive node now, you can use `geth` and I provide archive slices for it here: https://github.com/allada/bsc-archive-snapshot

In short... be patient, it will take a while.

**Comment by radian-rgc** [comment]
@allada i used before `Geth` service and currently i want to switch to `Erigon`, is it possible to set datadir to `Geth` node ?

**Comment by allada** [comment]
> @allada i used before `Geth` service and currently i want to switch to `Erigon`, is it possible to set datadir to `Geth` node ?

No, `geth`'s data format is significantly different than `erigon`. It would take a long time to explain here, but it would even be slower to read the archive node's data and write it to erigon's format than it would be to just sync normally (unless you could fit the entire geth data directory in ram).

> This snapshot is full Archive? if yes, until from witch block? and i see download speed is too slow (20m/s).

Yes it is syncing as an archive node from block 1. I have been uploading different versions as it progresses (use `aws s3 ls --request-payer=requester s3://public-blockchain-snapshots/bsc/erigon-` to see what is the latest), but I have been uploading as zfs snapshots instead of tar snapshots because it is easier for me, so you'll need to put it on a drive formatted with zfs as the filesystem.

**Comment by radian-rgc** [comment]
@allada Thank you so much,

Im cheking your last snapshot and i see its almost 500Gb. 
Im very intresting about ZSF 😄

Now i have a question about ZSF, 

Currently i have 2TB SSD Nvme, its enough for make Archive mode with ZSF?

Actually just i need the last 3Mil block with Archive and another can be without Archive...

Its possible?

**Comment by allada** [comment]
> Currently i have 2TB SSD Nvme, its enough for make Archive mode with ZSF?
I don't know, i have not finished a full sync yet.

> Actually just i need the last 3Mil block with Archive and another can be without Archive...
> 
> Its possible?
I am not sure, I'm fairly new to `Erigon`, but probably. It is possible with `geth`.

**Comment by radian-rgc** [comment]
> but I have been uploading as zfs snapshots instead of tar snapshots because it is easier for me, so you'll need to put it on a drive formatted with zfs as the filesystem.

I download the last snapshot (`erigon-partial-2021-12-25.zfs.zstd`) and i put it to drive formatted with zfs.

After this, I must decompress it or something ? I search about it but get nothing...

I have 2TB free with zfs drive formatted.

**Comment by allada** [comment]
Something like:
`aws s3 cp s3://path-to-data - | zstd -d | zfs recv datasetname`

You may want to enable lz4 compression first on the zfs dataset or it may
not fit in 2TB.

Keep in mind, it is not finished syncing. This is just a snapshot of the
current sync state.

**Comment by radian-rgc** [comment]
> Something like: `aws s3 cp s3://path-to-data - | zstd -d | zfs recv datasetname` 

Sorry but im using, but no any results...

`aws s3  --request-payer=requester cp s3://public-blockchain-snapshots/bsc/erigon-partial-2021-12-25.zfs.zstd - | zstd --long=31 -d | zfs recv dexset -F`

**Comment by allada** [comment]
Any output?

**Comment by radian-rgc** [comment]
No any output, i see runing in background, but no any making file/folder.

you can check it by ownNo any output, i see runing in background, but no any making file/folder.

you can check it by own

**Comment by allada** [comment]
> No any output, i see runing in background, but no any making file/folder.
> 
> you can check it by own

You can run `zfs list` to see if it is growing.
